cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java

```diff
@@ -0,0 +1,125 @@
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.util.Base64;
+
+/**
+ * Test for {@link FileFetcherHttpHandlerInternal}.
+ */
+public class FileFetcherHttpHandlerInternalTest extends AppFabricTestBase {
+  private static final Logger LOG = LoggerFactory.getLogger(FileFetcherHttpHandlerInternalTest.class);
+
+  private static DiscoveryServiceClient discoveryServiceClient;
+
+  @BeforeClass
+  public static void setup() {
+    discoveryServiceClient = getInjector().getInstance(DiscoveryServiceClient.class);
+  }
+
+  @AfterClass
+  public static void stop() {
+    AppFabricTestHelper.shutdown();
+  }
+
+  @Test
+  public void testDownloadFile() throws IOException {
+    // Create a jar file
+    LocalLocationFactory locationFactory = new LocalLocationFactory(tmpFolder.newFolder());
+    Location location = AppJarHelper.createDeploymentJar(locationFactory, AllProgramsApp.class);
+    File orgFile = new File(location.toURI().getPath());
+
+    // Set the target file path to store downloaded file content.
+    File targetFile = new File(tmpFolder.newFolder(), "target.jar");
+    Location targetLocation = new LocalLocationFactory().create(targetFile.toURI());
+
+    // Download file from AppFabric
+    RemoteClient remoteClient = new RemoteClient(
+      discoveryServiceClient,
+      Constants.Service.APP_FABRIC_HTTP,
+      new DefaultHttpRequestConfig(false), Constants.Gateway.INTERNAL_API_VERSION_3);
+    OutputStream outputStream = targetLocation.getOutputStream();
+
+    HttpRequest request = remoteClient.requestBuilder(
+      HttpMethod.GET,
+      String.format("location/%s", Base64.getEncoder().encodeToString(location.toURI().toString().getBytes())))
+      .withContentConsumer(
+      new HttpContentConsumer() {
+      @Override
+      public boolean onReceived(ByteBuffer chunk) {
+        try {
+          byte[] bytes = new byte[chunk.remaining()];
+          chunk.get(bytes, 0, bytes.length);
+          outputStream.write(bytes);
+          outputStream.flush();
+        } catch (IOException e) {
+          LOG.error("Failed to write to orgFile {}", targetLocation.toURI());
+          return false;
+        }
+        return true;
+      }
+
+      @Override
+      public void onFinished() {
+        try {
+        outputStream.close();
+        } catch (Exception e) {
+          LOG.error("Failed to close to orgFile {}", targetLocation.toURI());
+        }
+      }
+    }).build();
+    HttpResponse httpResponse = HttpRequests.execute(request, new DefaultHttpRequestConfig(false));
+    httpResponse.consumeContent();
+    Assert.assertEquals(httpResponse.getResponseCode(), HttpResponseStatus.OK.code());
+
+    // Verify the source and destination files are identical.
+    byte[] orgFileContent = Files.readAllBytes(orgFile.toPath());
+    byte[] downloadedFileContent = Files.readAllBytes(targetFile.toPath());
+    Assert.assertEquals(orgFileContent.length, downloadedFileContent.length);
```

Comment:
Just use `Assert.assertArrayEquals()` to unify to one assert call.

Reply:
Done. Fixed. 

I broke this up into length check and content check since it facilitated my debugging, as my earlier implementation incorrectly wrote the entire buffer to file, instead of the portion of the buffer containing data. The length check better surfaces that issue.